scripts/ghidra/PatchestryDecompileFunctions.java

```diff
@@ -156,6 +205,176 @@ private HighVariable lValueOf(Varnode node) throws Exception {
         	return var;
         }
 
+        private void serializePointerType(Pointer ptr) throws Exception {
+            name("name").value(ptr.getDisplayName());
+            name("kind").value("pointer");
+            name("size").value(ptr.getLength());
+            DataType elem_type = ptr.getDataType();
+            // Pointer element type could be null; assign void type to
+            // element in such case
+            if (elem_type == null) {
+                elem_type = VoidDataType.dataType;
+            }
+
+            name("element_type").value(label(elem_type));
+        }
+
+        private void serializeTypedefType(TypeDef typedef) throws Exception {
+            name("name").value(typedef.getDisplayName());
+            name("kind").value("typedef");
+            name("size").value(typedef.getLength());
+
+            DataType base_type = typedef.getBaseDataType();
+            // If base type is null, raise an exception.
+            // TODO(kumarak): if see cases where base type is null, handle it
+            // by assigning default `void` type.
+            if (base_type == null) {
+                throw new Exception("Typedef base type is null: " + typedef.toString());
+            }
+
+            name("base_type").value(label(base_type));
+        }
+
+        private void serializeArrayType(Array arr) throws Exception {
+            name("name").value(arr.getDisplayName());
+            name("kind").value("array");
+            name("size").value(arr.getLength());
+            name("num_elements").value(arr.getNumElements());
+            DataType elem_type = arr.getDataType();
+            if (elem_type != null) {
+                name("element_type").value(label(elem_type));
+            }
+        }
+            
+        private void serializeBuiltinType(DataType data_type, String kind) throws Exception {
+            name("name").value(data_type.getDisplayName());
+            name("size").value(data_type.getLength());
+            name("kind").value(kind);
+        }
+
+        private void serializeCompositeType(Composite data_type, String kind) throws Exception {
+            name("name").value(data_type.getDisplayName());
+            name("kind").value(kind);
+            name("size").value(data_type.getLength());
+            name("fields").beginArray();
+
+            for (int i = 0; i < data_type.getNumComponents(); i++) {
+                DataTypeComponent dtc = data_type.getComponent(i);
+                beginObject();
+
+                name("type").value(label(dtc.getDataType()));
+                name("offset").value(dtc.getOffset());
+
+                if (dtc.getFieldName() != null) {
+                    name("name").value(dtc.getFieldName().replaceAll(" ", ""));
```

Comment:
Why drop spaces? That might make two fields have the same name.

Reply:
Note: We don't need to produce valid C. Identifiers manually created via clang / mlir APIs can have whitespace.